lib/connection.js

```diff
@@ -818,32 +818,40 @@ Connection.prototype.dropCollection = async function dropCollection(collection)
 /**
  * Waits for connection to be established, so the connection has a `client`
  *
+ * @param {Boolean} [noTimeout=false] if set, don't put a timeout on the operation. Used internally so `mongoose.model()` doesn't leave open handles.
  * @return Promise
  * @api private
  */
 
-Connection.prototype._waitForConnect = async function _waitForConnect() {
+Connection.prototype._waitForConnect = async function _waitForConnect(noTimeout) {
   if ((this.readyState === STATES.connecting || this.readyState === STATES.disconnected) && this._shouldBufferCommands()) {
     const bufferTimeoutMS = this._getBufferTimeoutMS();
     let timeout = null;
     let timedOut = false;
     // The element that this function pushes onto `_queue`, stored to make it easy to remove later
     const queueElement = {};
-    await Promise.race([
-      new Promise(resolve => {
+    if (noTimeout) {
+      await new Promise(resolve => {
         queueElement.fn = resolve;
         this._queue.push(queueElement);
-      }),
-      new Promise(resolve => {
-        timeout = setTimeout(
-          () => {
-            timedOut = true;
-            resolve();
-          },
-          bufferTimeoutMS
-        );
-      })
-    ]);
+      });
```

Comment:
I think this could be de-duplicated by storing the promise in a variable before the if, and just directly awaiting it here and in the other branch adding it to the race.

Reply:
Fixed in https://github.com/Automattic/mongoose/pull/15251/commits/0a2da7115116fae133f8242965dcc58cc95549ff